nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/oidc/OidcService.java

```diff
@@ -0,0 +1,207 @@
+     *
+     * @return whether OpenId Connect is enabled
+     */
+    public boolean isOidcEnabled() {
+        return identityProvider.isOidcEnabled();
+    }
+
+    /**
+     * Returns the OpenId Connect authorization endpoint.
+     *
+     * @return the authorization endpoint
+     */
+    public URI getAuthorizationEndpoint() {
+        return identityProvider.getAuthorizationEndpoint();
+    }
+
+    /**
+     * Returns the OpenId Connect scope.
+     *
+     * @return scope
+     */
+    public Scope getScope() {
+        return identityProvider.getScope();
+    }
+
+    /**
+     * Returns the OpenId Connect client id.
+     *
+     * @return client id
+     */
+    public String getClientId() {
+        return identityProvider.getClientId().getValue();
+    }
+
+    /**
+     * Initiates an OpenId Connection authorization code flow using the specified request identifier to maintain state.
+     *
+     * @param oidcRequestIdentifier request identifier
+     * @return state
+     */
+    public State createState(final String oidcRequestIdentifier) {
+        if (!isOidcEnabled()) {
+            throw new IllegalStateException(OPEN_ID_CONNECT_SUPPORT_IS_NOT_CONFIGURED);
+        }
+
+        final CacheKey oidcRequestIdentifierKey = new CacheKey(oidcRequestIdentifier);
+        final State state = new State(new BigInteger(130, new SecureRandom()).toString(32));
+
+        try {
+            synchronized (stateLookupForPendingRequests) {
+                final State cachedState = stateLookupForPendingRequests.get(oidcRequestIdentifierKey, () -> state);
+                if (!state.equals(cachedState)) {
+                    throw new IllegalStateException("An existing login request is already in progress.");
+                }
+            }
+        } catch (ExecutionException e) {
+            throw new IllegalStateException("Unable to store the login request state.");
+        }
+
+        return state;
+    }
+
+    /**
+     * Validates the purposed state with the given request identifier. Will return false if the
+     * state does not match or if entry for this request identifier has expired.
+     *
+     * @param oidcRequestIdentifier request identifier
+     * @param purposedState purposed state
+     * @return whether the state is valid or not
+     */
+    public boolean isStateValid(final String oidcRequestIdentifier, final State purposedState) {
+        if (!isOidcEnabled()) {
+            throw new IllegalStateException(OPEN_ID_CONNECT_SUPPORT_IS_NOT_CONFIGURED);
+        }
+
+        final CacheKey oidcRequestIdentifierKey = new CacheKey(oidcRequestIdentifier);
+
+        synchronized (stateLookupForPendingRequests) {
+            final State state = stateLookupForPendingRequests.getIfPresent(oidcRequestIdentifierKey);
+            if (state != null) {
+                stateLookupForPendingRequests.invalidate(oidcRequestIdentifierKey);
+            }
+
+            return state != null && state.equals(purposedState);
```

Comment:
I looked at [`State#equals()`](https://www.javadoc.io/doc/com.nimbusds/oauth2-oidc-sdk/5.34) and it uses the standard Java `String#equals()` method which short-circuits on inequality. This leaks information and could lead to a timing attack vulnerability. We should use a [constant-time equality comparison](https://codahale.com/a-lesson-in-timing-attacks/) (i.e. iterate over every byte and use XOR, comparing the final result to `0`). One is readily available in [`MessageDigest.isEqual(byte[] a, byte[] b)`](https://docs.oracle.com/javase/8/docs/api/java/security/MessageDigest.html#isEqual-byte:A-byte:A-) or can be implemented and encapsulated ourselves in either the `CryptoUtils` or `CipherUtility` class.